### ISSUE TITLE
Syntax error in build_environment.py

### ISSUE THREAD

**Issue by ashkan2200** [LLNL/spack#2359]
Hi all,
I tried compiling `hdf5` but the build process crashes with the following error. Note that this crash happens after the dependencies are installed and `spack` wants to compiler `hdf5` itself

    spack/lib/spack/spack/build_environment.py:563, in child_execution:
         548      def child_execution(child_connection, input_stream):
         549          try:
         550              setup_package(pkg, dirty=dirty)
         551              function(input_stream)
         552              child_connection.send(None)
         553          except:
         554              # catch ANYTHING that goes wrong in the child process
         555              exc_type, exc, tb = sys.exc_info()
         556  
         557              # Need to unwind the traceback in the child because traceback
         558              # objects can't be sent to the parent.
         559              tb_string = traceback.format_exc()
         560  
         561              # build up some context from the offending package so we can
         562              # show that, too.
      >> 563              package_context = get_package_context(tb)
         564  
         565              build_log = None
         566              if hasattr(pkg, 'log_path'):
         567                  build_log = pkg.log_path
         568  
         569              # make a pickleable exception to send to parent.
         570              msg = "%s: %s" % (str(exc_type.__name__), str(exc))
         571  
         572              ce = ChildError(msg, tb_string, build_log, package_context)
         573              child_connection.send(ce)
         574  
         575          finally:
         576              child_connection.close()

**Comment by adamjstewart** [comment]
Are you on macOS? I was able to reproduce your error message, but it didn't have anything to do with HDF5 for me. See #2363 for details.

**Comment by ashkan2200** [comment]
I tried it on scientific linux.

**Comment by adamjstewart** [comment]
HDF5 built fine for me on CentOS 6 on the latest develop. I'm not sure what's going on. Maybe @alalazo would have some insight.

**Comment by alalazo** [comment]
@adamjstewart Cannot reproduce. Just built `hdf5` on the lastest `develop`:
```console
[+] /home/mculpo/wdir/spack/opt/spack/linux-ubuntu14-x86_64/gcc-6.2.0/hdf5-1.10.0-patch1-46gi64flrwsvwhvemtonf2fuocgqupz4

$ spack spec -Il hdf5
Input spec
--------------------------------
     mscv5vo  hdf5

Normalized
--------------------------------
     uqxsqv5  hdf5
     j52z7qb      ^zlib@1.1.2:

Concretized
--------------------------------
[+]  46gi64f  hdf5@1.10.0-patch1%gcc@6.2.0+cxx~debug+fortran+mpi+shared~szip~threadsafe arch=linux-ubuntu14-x86_64
[+]  dayzkif      ^openmpi@2.0.1%gcc@6.2.0~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs+vt arch=linux-ubuntu14-x86_64
[+]  mt7dzme          ^hwloc@1.11.4%gcc@6.2.0 arch=linux-ubuntu14-x86_64
[+]  l7oxfng              ^libpciaccess@0.13.4%gcc@6.2.0 arch=linux-ubuntu14-x86_64
[+]  g3ilmj6                  ^libtool@2.4.6%gcc@6.2.0 arch=linux-ubuntu14-x86_64
[+]  buhbb2u                      ^m4@1.4.17%gcc@6.2.0+sigsegv arch=linux-ubuntu14-x86_64
[+]  zvh4aiz                          ^libsigsegv@2.10%gcc@6.2.0 arch=linux-ubuntu14-x86_64
[+]  pqdron5                  ^pkg-config@0.29.1%gcc@6.2.0+internal_glib arch=linux-ubuntu14-x86_64
[+]  qcgxbkc                  ^util-macros@1.19.0%gcc@6.2.0 arch=linux-ubuntu14-x86_64
[+]  hahiy3j      ^zlib@1.2.8%gcc@6.2.0 arch=linux-ubuntu14-x86_64
```

**Comment by ashkan2200** [comment]
I checked a bit more and this issue happens when I use external `openmpi`

**Comment by pramodk** [comment]
I am trying Spack on Cray KNL system (discussion in #2572) and seeing exactly same issue when I try to install :

```
$ spack spec pdt %intel
Input spec
--------------------------------
pdt%intel
Normalized
--------------------------------
pdt%intel
Concretized
--------------------------------
pdt@3.22.1%intel@17.0.1.132 arch=cray-CNL-mic_knl

$ spack install -v pdt %intel
==> Installing pdt
==> Error: SyntaxError: invalid syntax (<string>, line 502)
/xxxxx/home/kumbhar/spack/lib/spack/spack/build_environment.py:563, in child_execution:
     548      def child_execution(child_connection, input_stream):
     549          try:
     550              setup_package(pkg, dirty=dirty)
     551              function(input_stream)
     552              child_connection.send(None)
     553          except:
     554              # catch ANYTHING that goes wrong in the child process
     555              exc_type, exc, tb = sys.exc_info()
     556
     557              # Need to unwind the traceback in the child because traceback
     558              # objects can't be sent to the parent.
     559              tb_string = traceback.format_exc()
     560
     561              # build up some context from the offending package so we can
     562              # show that, too.
  >> 563              package_context = get_package_context(tb)
```
(I don't have this issue on os-x, x86, bg-q platforms.)

**Comment by Exteris** [comment]
I am getting the same issue with external openmpi, on CentOS 7 (Marconi), with symlinked modulecmd -> modulecmd.tcl

```
==> Error: SyntaxError: invalid syntax (<string>, line 1)
/marconi/home/userexternal/dvanvugt/spack/lib/spack/spack/build_environment.py:567, in child_execution:
```

**Comment by becker33** [comment]
@ashkan2200 is this issue occurring on a Cray machine or some other platform?

**Comment by ashkan2200** [comment]
@becker33 , no, the issue was on a cluster with infiniband network.

**Comment by becker33** [comment]
@ashkan2200 Is this resolved now (as of #3250 being merged)?

**Comment by becker33** [comment]
Closing as fixed by #3250.
